interfaces/repo.go

```diff
@@ -516,3 +516,110 @@ func (r *Repository) securitySnippetsForSnap(snapName string, securitySystem Sec
 	}
 	return snippets, nil
 }
+
+// AddSnap adds plugs and slots declared by the given snap to the repository.
+//
+// This function can be used to implement snap install or, when used along with
+// RemoveSnap, snap upgrade.
+//
+// Plugs/slots that are already present in the repository are not changed. Only
+// new plugs/slots are added.
+//
+// Each added plug/slot is validated according to the corresponding interface.
+// Unknown interfaces and plugs/slots that don't validate are not added.
+// Information about those failures are returned to the caller.
+func (r *Repository) AddSnap(snapInfo *snap.Info) []error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	snapName := snapInfo.Name
+
+	for plugName := range snapInfo.Plugs {
+		if _, ok := r.plugs[snapName][plugName]; ok {
+			panic(fmt.Errorf("cannot add plug %s.%s: existing plug in the way", snapName, plugName))
```

Comment:
This shouldn't be a panic. The error happens as a side effect of state changed in a place completely unrelated to the call site, which means it's very easy to do that mistake by accident rather than obviously incorrect code. If that ever happens, what do we prefer: an error message to the user, or a non-working system?